05-complexity-determining-complexity/determining_complexity_answers.md

```diff
@@ -0,0 +1,291 @@
+ min = from
+ max = to
+
+ # Current free slot
+ free = min
+
+ while min < max
+   if free == min # Evaluate collection[max]
+     if collection[max] <= pivot # Smaller than pivot, must move
+       collection[free] = collection[max]
+       min += 1
+       free = max
+     else
+       max -= 1
+     end
+   elsif free == max # Evaluate collection[min]
+     if collection[min] >= pivot # Bigger than pivot, must move
+       collection[free] = collection[min]
+       max -= 1
+       free = min
+     else
+       min += 1
+     end
+   else
+     raise "Inconsistent state"
+   end
+ end
+
+ collection[free] = pivot
+
+ quick_sort collection, from, free - 1
+ quick_sort collection, free + 1, to
+
+ collection
+end
+```
+T(n): Time taken to calculate quick_sort
+T(0) = T(1) = 1
+
+<Worst-case: When pivot doesn't partition evenly, which means pivot is the least or the largest element, subcollection is only 1 less than the collection. This recursive call takes C(n-1) time and the next recursive call takes C(n-2) time and so on.>
+
+*(n-1)C: time taken to do while loop*
+*T(n-1): time taken to recursively call on the most unbalanced subcollection*
+*1: time taken to recursively call on the empty array. After partitioning, two subcollection sizes are 0 and n-1*
+
+T(n) = (n-1)C + T(n-1) + 1    # T(n-1) = (n-2)C + T(n-2) + 1
+     = (n-1)C + {(n-2)C + T(n-2) + 1} + 1
+     = 2Cn - (1+2)C + T(n-2) + 2   # T(n-2) = (n-3)C + T(n-3) + 1
+     = 2Cn - (1+2)C + {(n-3)C + T(n-3) + 1} + 2
+     = 3Cn - (1+2+3)C + T(n-3) + 3    # T(n-3) = (n-4)C + T(n-4) + 1
+     = 3Cn - (1+2+3)C + {(n-4)C + T(n-4) + 1} + 3
+     = 4Cn - (1+2+3+4)C + T(n-4) + 4
+     ...
+     = kCn - C(1+k)k/2 + T(n-k) + k    # if n-k = 0, n = k
+     = Cn^2 - C(1+n)n/2 + T(0) + n
+     = Cn^2 - Cn/2 - C/2 n^2 + 1 + n
+     = (C - C/2)n^2 + (1 - C/2)n + 1
+
+
+     => O(n^2)   # drops constant coefficient, lower order term and constant term
+
+
+<Best-case: When pivot is in the middle after partitioning, two subcollection sizes are equal or within 1 of each other.>
+*(n-1)C: time taken to do while loop*
+*T(n/2): time taken to recursively call on the evenly partitioned subcollection*
+*After partitioning, two subcollection sizes are n/2 and n/2*
+
+T(n) = (n-1)C + T(n/2) + T(n/2)
+     = (n-1)C + 2T(n/2)   # T(n/2) = (n/2 - 1)C + 2T(n/4)
+     = (n-1)C + 2{ (n/2 - 1)C + 2T(n/4) }
+     = 2Cn - 3C + 4T(n/4)   # T(n/4) = (n/4 - 1)C + 2T(n/8)
+     = 2Cn - 3C + 4{ (n/4 - 1)C + 2T(n/8) }
+     = 3Cn - 7C + 8T(n/8)   # T(n/8) = (n/8 - 1)C + 2T(n/16)
+     = 3Cn - 7C + 8{ (n/8 - 1)C + 2T(n/16) }
+     = 4Cn - 15C + 16T(n/16)
+     ...
+     = kCn - (2^k -1)C + 2^k T(n/2^k)   # if n/2^k = 1, n = 2^k. So k = lg n
+     = Cn lg n - (2^lg n - 1)C + 2^lg n T(1)   # 2^lg n = n
+     = Cn lg n - Cn + C + n
+     = Cn lg n + (1 - C)n + C
+
+
+     => O(n lg n) = O(n log n)  # drops constant coefficient, lower order term and constant term
+
+
+Worst-case: O(n^2)
+Best-case: O(n log n)
```

Comment:
When talking about the best case, you'll want to use Omega (Ω), so it would be Ω(n log n).